src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -1333,6 +1345,585 @@ bool ReplicatedWriteLog<I>::alloc_resources(C_BlockIORequestT *req) {
+   * finish flushing.
+   */
+
+  if (m_flush_ops_in_flight &&
+      (log_entry->ram_entry.sync_gen_number > m_lowest_flushing_sync_gen)) {
+    return false;
+  }
+
+  return (log_entry->can_flush() &&
+         (m_flush_ops_in_flight <= IN_FLIGHT_FLUSH_WRITE_LIMIT) &&
+         (m_flush_bytes_in_flight <= IN_FLIGHT_FLUSH_BYTES_LIMIT));
+}
+
+template <typename I>
+Context* ReplicatedWriteLog<I>::construct_flush_entry_ctx(std::shared_ptr<GenericLogEntry> log_entry) {
+  //TODO handle writesame, invalidate and discard in later PRs
+  CephContext *cct = m_image_ctx.cct;
+  bufferlist entry_bl;
+
+  ldout(cct, 20) << "" << dendl;
+  ceph_assert(m_entry_reader_lock.is_locked());
+  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+  if (!m_flush_ops_in_flight ||
+      (log_entry->ram_entry.sync_gen_number < m_lowest_flushing_sync_gen)) {
+    m_lowest_flushing_sync_gen = log_entry->ram_entry.sync_gen_number;
+  }
+  auto gen_write_entry = static_pointer_cast<GenericWriteLogEntry>(log_entry);
+  m_flush_ops_in_flight += 1;
+  /* For write same this is the bytes affected bt the flush op, not the bytes transferred */
+  m_flush_bytes_in_flight += gen_write_entry->ram_entry.write_bytes;
+
+  gen_write_entry->flushing = true;
+
+  auto write_entry = static_pointer_cast<WriteLogEntry>(log_entry);
+  /* Pass a copy of the pmem buffer to ImageWriteback (which may hang on to the bl evcen after flush()). */
+  buffer::list entry_bl_copy;
+  write_entry->copy_pmem_bl(&entry_bl_copy);
+  entry_bl_copy.begin(0).copy(write_entry->write_bytes(), entry_bl);
+
+  /* Flush write completion action */
+  Context *ctx = new LambdaContext(
+    [this, gen_write_entry](int r) {
+      {
+        std::lock_guard locker(m_lock);
+        gen_write_entry->flushing = false;
+        if (r < 0) {
+          lderr(m_image_ctx.cct) << "failed to flush log entry"
+                                 << cpp_strerror(r) << dendl;
+          m_dirty_log_entries.push_front(gen_write_entry);
+        } else {
+          ceph_assert(!gen_write_entry->flushed);
+          gen_write_entry->flushed = true;
+          ceph_assert(m_bytes_dirty >= gen_write_entry->bytes_dirty());
+          m_bytes_dirty -= gen_write_entry->bytes_dirty();
+          sync_point_writer_flushed(gen_write_entry->sync_point_entry);
+          ldout(m_image_ctx.cct, 20) << "flushed: " << gen_write_entry
+                                     << dendl;
+        }
+        m_flush_ops_in_flight -= 1;
+        m_flush_bytes_in_flight -= gen_write_entry->ram_entry.write_bytes;
+        wake_up();
+      }
+    });
+  /* Flush through lower cache before completing */
+  ctx = new LambdaContext(
+    [this, ctx](int r) {
+      if (r < 0) {
+        lderr(m_image_ctx.cct) << "failed to flush log entry"
+                               << cpp_strerror(r) << dendl;
+        ctx->complete(r);
+      } else {
+        m_image_writeback.aio_flush(ctx);
+      }
+    });
+
+  if (log_entry->is_write()) {
```

Comment:
@dillaman  We need is_write() here. If we remove is_write() here, we need to put m_image_writeback as a parameter in xxxLogEntry class. Then we need a virtual function member template. However, xxxLogEntry class is not a template. A member function template cannot be virtual, and a member function template in a derived class cannot override a virtual member function from the base class.

Reply:
@dillaman added librbd::cache::ImageWritebackInterface.